src/components/upload-modal/__tests__/UploadModal.test.tsx

```diff
@@ -0,0 +1,61 @@
+import { fireEvent, render, screen } from '@testing-library/react'
+import { describe, it, expect, vi } from 'vitest'
+import UploadModal from '..'
+import { Button } from '../../ui/button'
+
+describe('UploadModal', () => {
+  it('should render correctly', () => {
+    render(<UploadModal onUpload={() => {}} onDelete={() => {}} uploaded={''} />)
+
+    expect(screen.getByText('Upload')).toBeDefined()
+  })
+
+  it('should render custom button', () => {
+    render(<UploadModal onUpload={() => {}} onDelete={() => {}} uploaded={''} customButton={<Button>Custom Button</Button>} />)
+
+    expect(screen.getByText('Custom Button')).toBeDefined()
+  })
+
+  it('should open modal when button is clicked', () => {
+    render(<UploadModal onUpload={() => {}} onDelete={() => {}} uploaded={''} />)
+
+    fireEvent.click(screen.getByText('Upload'))
+
+    expect(screen.getByText('Upload a file for the track')).toBeDefined()
+  })
+
+  it('should accept valid MP3 file', () => {
+    render(<UploadModal onUpload={() => {}} onDelete={() => {}} uploaded="" />)
+    
+    fireEvent.click(screen.getByText('Upload'))
+    
+    const file = new File(['audio content'], 'test.mp3', { type: 'audio/mpeg' })
+    const input = screen.getByTestId('file-input')
+    
+    fireEvent.change(input, { target: { files: [file] } })
+    
+    expect(screen.getByText('Selected: test.mp3')).toBeDefined()
+  })
+
+  it('should reject invalid file types', async () => {
+    render(<UploadModal onUpload={() => {}} onDelete={() => {}} uploaded="" />)
+
+    fireEvent.click(screen.getByText('Upload'))
+
+    const file = new File(['audio content'], 'test.txt', { type: 'text/plain' })
+    const input = screen.getByTestId('file-input')
+
+    fireEvent.change(input, { target: { files: [file] } })
+    fireEvent.click(screen.getByTestId('upload-button'))
+
+    await expect(screen.findByText('File must be a valid MP3 or WAV file')).resolves.toBeDefined()
```

Comment:
Might be a good idea to check, that the file was not selected

Reply:
The validation happens only when user clicks "Upload" button, so the file is actually selected here. Don't want to change how it works, in this PR at least, so I didn't add the check.